tsl/src/remote/async.c

```diff
@@ -512,6 +512,137 @@ async_request_wait_ok_result(AsyncRequest *req)
 	return res;
 }
 
+/*
+ * Get the result of an async request during cleanup.
+ *
+ * Cleanup is typically necessary for a query that is being interrupted by
+ * transaction abort, or a query that was initiated as part of transaction
+ * abort to get the remote side back to the appropriate state.
+ *
+ * endtime is the time at which we should give up and assume the remote
+ * side is dead.
+ *
+ * An AsyncReponse is always returned, indicating last PGresult received,
+ * a timeout, or error.
+ */
+AsyncResponse *
+async_request_cleanup_result(AsyncRequest *req, TimestampTz endtime)
+{
+	volatile bool timed_out = false;
+	volatile bool conn_issue = false;
+	PGresult *volatile last_res = NULL;
+	const TSConnection *conn = async_request_get_connection(req);
+	PGconn *pg_conn = remote_connection_get_pg_conn(conn);
+
+	switch (req->state)
+	{
+		case DEFERRED:
+			if (remote_connection_is_processing(req->conn))
+				return async_response_error_create("request already in progress");
+
+			req = async_request_send_internal(req, WARNING);
+
+			if (req == NULL)
+				return async_response_error_create("failed to send deferred request");
+
+			Assert(req->state == EXECUTING);
+			break;
+		case EXECUTING:
+			break;
+		case COMPLETED:
+			return async_response_error_create("request already completed");
+	}
+
+	/* In what follows, do not leak any PGresults on an error. */
+	PG_TRY();
+	{
+		for (;;)
+		{
+			PGresult *res;
+
+			while (PQisBusy(pg_conn))
+			{
+				int wc;
+				TimestampTz now = GetCurrentTimestamp();
+				long secs;
+				int microsecs;
+				long cur_timeout;
+
+				/* If timeout has expired, give up, else get sleep time. */
+				if (now >= endtime)
+				{
+					timed_out = true;
+					goto exit;
+				}
+				TimestampDifference(now, endtime, &secs, &microsecs);
+
+				/* To protect against clock skew, limit sleep to one minute. */
+				cur_timeout = Min(60000, secs * USECS_PER_SEC + microsecs);
+
+				/* Sleep until there's something to do */
+				wc = WaitLatchOrSocket(MyLatch,
+									   WL_LATCH_SET | WL_SOCKET_READABLE
+#if PG12_GE
+										   | WL_EXIT_ON_PM_DEATH
+#endif
+										   | WL_TIMEOUT,
+									   PQsocket(pg_conn),
+									   cur_timeout,
+									   PG_WAIT_EXTENSION);
+				ResetLatch(MyLatch);
+
+				CHECK_FOR_INTERRUPTS();
+
+				/* Data available in socket? */
+				if (wc & WL_SOCKET_READABLE)
+				{
+					if (0 == PQconsumeInput(pg_conn))
+					{
+						conn_issue = true;
+						goto exit;
+					}
+				}
+			}
+
+			res = PQgetResult(pg_conn);
+
+			if (res == NULL)
+			{
+				remote_connection_set_processing(req->conn, false);
+				async_request_set_state(req, COMPLETED);
+				break; /* query is complete */
+			}
+
+			PQclear(last_res);
+			last_res = res;
+		}
+	exit:;
```

Comment:
Goto labels are usually without any indentation. Should this place be consistent too?

Reply:
This was run through formatting, and I guess it is indented to show that is inside the context of the try-catch braces.

Reply:
Can it be due to `;` on the same line as the label?

Reply:
The semicolon is necessary because otherwise there's a parser error given that there is no code before the closing brace. This is the way the code was when importing it from postgresql.

Reply:
I understand that it is necessary, sorry I wasn't clear: you can move it to another line, so they are not on the same line?

Reply:
Again, the formatter prevents that. It removes the extra newline so moving the `;` to a new line doesn't make a difference. FWIW, I don't think it is wrong to have the goto tag at the "right" indentation level. Besides, this is what the formatter gives us.